app/controllers/legacy/sessions_controller.rb

```diff
@@ -0,0 +1,305 @@
+            redirect_to action: :redirect_back
+          when :new_social_user
+            security_log :sign_in_successful, authentication_id: authentication.id
+            redirect_to signup_profile_path
+          when :authentication_added
+            security_log :authentication_created,
+                        authentication_id: authentication.id,
+                        authentication_provider: authentication.provider,
+                        authentication_uid: authentication.uid
+            security_log :sign_in_successful, authentication_id: authentication.id
+            redirect_to profile_path,
+                        notice: I18n.t(:"controllers.sessions.new_sign_in_option_added")
+          when :authentication_taken
+            security_log :authentication_transfer_failed, authentication_id: authentication.id
+            redirect_to profile_path,
+                        alert: I18n.t(:"controllers.sessions.sign_in_option_already_used")
+          when :same_provider
+            security_log :authentication_transfer_failed, authentication_id: authentication.id
+            redirect_to profile_path,
+                        alert: I18n.t(:"controllers.sessions.same_provider_already_linked",
+                                      user_name: current_user.name,
+                                      authentication: authentication.display_name)
+          when :mismatched_authentication
+            security_log :sign_in_failed, reason: "mismatched authentication"
+            redirect_to authenticate_path,
+                        alert: I18n.t(:"controllers.sessions.mismatched_authentication")
+          when :email_already_in_use
+            redirect_to profile_path,
+                        alert: I18n.t(:"controllers.sessions.way_to_login_cannot_be_added")
+          else
+            oauth = request.env['omniauth.auth']
+            errors = @handler_result.errors.inspect
+            last_exception = $!.inspect
+            exception_backtrace = $@.inspect
+
+            error_message = "[SessionsCreate] IllegalState on success: " +
+                            "OAuth data: #{oauth}; status: #{status}; " +
+                            "errors: #{errors}; last exception: #{last_exception}; " +
+                            "exception backtrace: #{exception_backtrace}"
+
+            # This will print the exception to the logs and send devs an exception email
+            raise IllegalState, error_message
+          end
+        end,
+        failure: -> do
+          errors = @handler_result.errors
+          lost_user = errors.any? do |error|
+            [:unknown_callback_state, :invalid_omniauth_data].include? error.code
+          end
+
+          if lost_user
+            # Something weird happened to the user's session or omniauth data and we lost their info
+            Rails.logger.warn do
+              oauth = request.env['omniauth.auth']
```

Comment:
this is far far too large a method.  difficult to understand, test, extend.

Reply:
Totally agree! 😅  It caused me a significant amount of headaches in the beginning when trying to understand Accounts. But then again, this is now legacy code that is already getting replaced entirely by new code, so not worth refactoring this legacy code IMO.